### PR TITLE
llama.cpp : split llama_context_params into model and context params

### PR DESCRIPTION
**Pull request by slaren**
Currently, `llama_load_model_from_file` and `llama_new_context_with_model` receive the same parameters, but it is not specified which parameters are relevant for each object. This change splits the parameters into `llama_model_params` and `llama_context_params` so that it is clear what parameters can be modified per-context.

Additionally:
- Uses the model `n_ctx_train` as the default context size when `llama_context_params::n_ctx == 0` or `-c 0`
- Moves `n_threads` to `llama_context_params`, adds `n_threads_batch` and the corresponding command line argument `-tb, --threads-batch` for batches/prompt processing
- Removes the `low_vram` parameter and the corresponding `-lv, --low-vram` command line argument

### CODE REVIEW

**Comment by slaren** [comment]
Despite what the prints say, it looks like the KV cache is offloaded regardless of the value of `low_vram`. Am I missing something here, or did this functionality get lost somewhere along the way? Tagging @JohannesGaessler.

https://github.com/ggerganov/llama.cpp/blob/bc9d3e3971e5607a10ff4c24e39568ce1ac87271/llama.cpp#L2323-L2338

https://github.com/ggerganov/llama.cpp/blob/bc9d3e3971e5607a10ff4c24e39568ce1ac87271/llama.cpp#L1229-L1236

**Comment by JohannesGaessler** [comment]
Both the `--low-vram` option and KV cache offloading were added in https://github.com/ggerganov/llama.cpp/pull/1827 . It seems that the program logic and the prints were inconsistent from the start. On master the results are still correct when combining `--low-vram` with KV cache offloading. However, the performance degrades relative to offloading only the repeating and non-repeating layers due to a much higher number of CPU<->GPU copies. I don't remember for certain what I did several months ago but I think the reason the prints say that the combination of `--low-vram` and KV cache offloading is not allowed is because supporting the combination would not make sense in terms of performance and I wanted to reduce the work needed for implementation/maintenance.

I'm not 100% sure how to proceed. The results happen to be correct, but as I said: supporting the combination does not make sense and I would not consider it to be in-scope. So I personally would still prefer to disallow it and to adjust the program logic to match the prints.

**Comment by slaren** [comment]
The reason this is relevant to this PR is that currently I have duplicated `low_vram` in `llama_model_params` and `llama_context_params`, since this paramter it is supposed to be used both when loading the model and when creating the KV cache. I don't think this is good, and I would prefer to remove it from `llama_context_params`. It would still be possible to do this and to match the logic of the prints by storing the value of `low_vram` in `llama_model`, but I am not convinced that it is worth to do that, since the same behavior can be achieved simply by using a smaller value of `n_gpu_layers`.

**Comment by JohannesGaessler** [comment]
Alright. I do not feel strongly about this and my line of reasoning is revolving around what is easier to implement and maintain anyways (assuming we want to keep `--low-vram` that would be to disallow the combination with KV cache offloading). Intuitively I would agree that thanks to your improvements regarding VRAM allocation (and performance optimizations in general) the use of `--low-vram` probably no longer makes sense and I would be fine with just removing the option. I cannot 100% guarantee that there isn't still some use case where `--low-vram` helps but if there are any they should be very rare.

**Comment by slaren** [comment]
I ran some tests with `low_vram`, and besides some unintended consequences of disabling the scratch buffer while still offloading the KV (some mat muls during attention wouldn't be offloaded anymore), I couldn't observe any meaningful differences in VRAM usage. After removing the effect of `low_vram` on the VRAM scratch buffer, the only remaining effect was preventing the offloading of the output norm weight, which is tiny anyway, just 16 KB on 7B. So I went ahead and removed the `low_vram` option entirely.

**Review by ggerganov — APPROVED** [review]
We can merge this together with #3228 to make the API change in one go

**Comment by netrunnereve** [comment]
Thanks for completing #2534! It's running with no issues on my system.

Note there are a few things from my PR that you may want to add here, such as README updates and including the thread count in the system info line.

**Comment by slaren** [comment]
While we are changing the API, should we also remove all the `_from_model` functions, and make the regular functions take a `llama_model` instead? I have already added a `llama_get_model` API to obtain the model of a context, so that users that only have a `llama_context` for whatever reason can still use these functions.

**Comment by ggerganov** [comment]
> While we are changing the API, should we also remove all the _from_model functions

Do you mean `_with_model`?
Yes, we should do that. Lets resolve conflicts and merge

**Comment by slaren** [comment]
Now that #3228 is merged, I will try to get this ready to merge as soon as possible. @ggerganov do you agree with removing the `_with_model` APIs as explained in my previous comment?

File: llama.cpp [hunk 1]
```diff
@@ -6606,6 +6560,24 @@ struct llama_context_params llama_context_default_params() {
     return result;
 }
 
+struct llama_context_params llama_context_default_params() {
+    struct llama_context_params result = {
+        /*.seed                        =*/ LLAMA_DEFAULT_SEED,
+        /*.n_ctx                       =*/ 512,
+        /*.n_batch                     =*/ 512,
+        /*.n_threads                   =*/ GGML_DEFAULT_N_THREADS, // TODO: better default
+        /*.n_threads_batch             =*/ GGML_DEFAULT_N_THREADS,
```
Comment (slaren) [inline llama.cpp:6569]: Should we change this to something like `std::thread::hardware_concurrency()`, or `get_num_physical_cores()` from `common.cpp`? It shouldn't affect most of our examples, but a better default could help downstream users.

**Comment by slaren** [comment]
If the API changes look good I think this can be merged.

**Pull request merged**